megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java

```diff
@@ -0,0 +1,1204 @@
+            .map(UnitState::position)
+            .filter(c -> c.distance(position) <= closingDistance)
+            .count();
+        double friendsCrowdingTolerance = antiCrowdingFactor * crowdingFriends;
+        double enemiesCrowdingTolerance = antiCrowdingFactor * crowdingEnemies;
+        return friendsCrowdingTolerance + enemiesCrowdingTolerance;
+    }
+
+    // P4
+    private static double calculateMovementMod(UnitAction unitAction, BehaviorParameters behaviorParameters) {
+        var tmmFactor = behaviorParameters.p4();
+        var tmm = Compute.getTargetMovementModifier(unitAction.hexesMoved(), unitAction.jumping(), false, null);
+        var tmmValue = clamp01(tmm.getValue() / 8.0);
+        return tmmValue * tmmFactor;
+    }
+
+    // P3
+    private static double getBraveryMod(UnitAction unitAction, Map<Integer, UnitState> currentUnitStates, BehaviorParameters behaviorParameters) {
+        var closestEnemy = closestNonAlliedEntity(unitAction, currentUnitStates);
+        var distanceToClosestEnemy = distanceToClosestNonAlliedUnit(unitAction, currentUnitStates);
+        int damageTaken = getDamageAtRange(distanceToClosestEnemy, closestEnemy);
+        int damageCaused = getDamageAtRange(distanceToClosestEnemy, currentUnitStates.get(unitAction.id()));
+        double successProbability = 1d - unitAction.chanceOfFailure();
+        return clamp01(0.1 + clamp01((successProbability * damageCaused * behaviorParameters.p3()) - damageTaken));
+    }
+
+    // P2
+    private static double calculateFallMod(UnitAction unitAction, BehaviorParameters behaviorParameters) {
+        double pilotingFailure = unitAction.chanceOfFailure();
+        double fallShameFactor = behaviorParameters.p2();
+        return clamp01((1 - fallShameFactor) + (1 - pilotingFailure) * fallShameFactor);
+    }
+
+    private static double calculateNearbyUnitsMod(UnitAction unitAction, Map<Integer, UnitState> currentUnitStates, BehaviorParameters behaviorParameters) {
+        double weight = behaviorParameters.p6();
+        double distance = distanceToClosestNonAlliedUnit(unitAction, currentUnitStates);
+        return clamp01(1.1 - weight + (1.0 / (distance + 1)) * weight);
+    }
+
+    private static double calculateHealthMod(UnitAction unitAction, BehaviorParameters behaviorParameters) {
+        double weight = behaviorParameters.p6();
+        double health = (unitAction.armorP() + unitAction.internalP()) / 2;
+        if (health < 0.7) {
+            weight *= 1.5; // More cautious when damaged
+        }
+        return health * weight;
+    }
+
+    private double calculateSelfPreservationMod(UnitAction unitAction, Map<Integer, UnitState> currentUnitStates, BehaviorParameters behaviorParameters) {
+        double weight = behaviorParameters.p6();
+        double health = (unitAction.armorP() + unitAction.internalP()) / 2;
+        if (health < 0.7) {
+            weight *= 1.5; // More cautious when damaged
+        }
```

Comment:
Is there a reason for this to be discrete rather than, say, using the lerp function to produce a continuous range?

Reply:
No, there is none, I reached a point where I started to throw things at the wall to see what sticks in the hope that I can make the loss of the cost function reach low enough.